api/include/opentelemetry/trace/propagation/b3_propagator.h

```diff
@@ -0,0 +1,278 @@
+#include "opentelemetry/trace/default_span.h"
+#include "opentelemetry/trace/propagation/http_text_format.h"
+#include "opentelemetry/trace/span.h"
+#include "opentelemetry/trace/span_context.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace trace
+{
+namespace propagation
+{
+
+static const nostd::string_view B3CombinedHeader = "b3";
+
+static const nostd::string_view B3TraceIdHeader = "X-B3-TraceId";
+static const nostd::string_view B3SpanIdHeader  = "X-B3-SpanId";
+static const nostd::string_view B3SampledHeader = "X-B3-Sampled";
+
+/*
+     B3, single header:
+                   b3: 80f198ee56343ba864fe8b2a57d3eff7-e457b5a2e4d86bd1-1-05e3ac9a4f6e3b90
+                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^ ^^^^^^^^^^^^^^^^
+                       0          TraceId            31 33  SpanId    48 | 52 ParentSpanId 68
+                                                                        50 Debug flag
+     Multiheader version:                                           X-B3-Sampled
+                             X-B3-TraceId                X-B3-SpanId    X-B3-ParentSpanId (ignored)
+*/
+
+static const int kTraceIdHexStrLength   = 32;
+static const int kSpanIdHexStrLength    = 16;
+static const int kTraceFlagHexStrLength = 1;
+
+// The B3PropagatorExtractor class provides an interface that enables extracting context from
+// headers of HTTP requests. HTTP frameworks and clients can integrate with B3Propagator by
+// providing the object  containing the headers, and a getter function for the extraction.
+// Based on:
+// https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/context/api-propagators.md#b3-extract
+template <typename T>
+class B3PropagatorExtractor : public HTTPTextFormat<T>
+{
+public:
+  // Rules that manages how context will be extracted from carrier.
+  using Getter = nostd::string_view (*)(const T &carrier, nostd::string_view trace_type);
+
+  // Returns the context that is stored in the HTTP header carrier with the getter as extractor.
+  context::Context Extract(Getter getter,
+                           const T &carrier,
+                           context::Context &context) noexcept override
+  {
+    SpanContext span_context = ExtractImpl(getter, carrier);
+    nostd::shared_ptr<Span> sp{new DefaultSpan(span_context)};
+    return context.SetValue(kSpanKey, sp);
+  }
+
+  static SpanContext GetCurrentSpan(const context::Context &context)
+  {
+    context::Context ctx(context);
+    context::ContextValue span = ctx.GetValue(kSpanKey);
+    if (nostd::holds_alternative<nostd::shared_ptr<Span>>(span))
+    {
+      return nostd::get<nostd::shared_ptr<Span>>(span).get()->GetContext();
+    }
+    return SpanContext::GetInvalid();
+  }
+
+  static TraceId GenerateTraceIdFromString(nostd::string_view trace_id)
+  {
+    uint8_t buf[kTraceIdHexStrLength / 2];
+    uint8_t *b_ptr = buf;
+    GenerateHexFromString(trace_id, kTraceIdHexStrLength, b_ptr);
```

Comment:
As B3 trace_id could also be 16 hex chars, and we are always trying to convert 32 hex chars. Should there be any check to avoid that, or else put some comment that it can fail.

Reply:
Yes! Thanks for pointing this out. I'll create fix and test for such case.